Support processing JS produced by transformers.

jods4 opened the pull request:
Fixes #217 

This PR adds a new callback parameter to transformers: `transpile`. 
When set to true by transformer, `bundleItem` is then considered as a normal script rather than just contents. This means that the bundler will process its dependencies and include them, so you can `require()` to your heart's contents.

There is no compatibility issue since this is a new parameter and when it's not set the code behaves exactly like before.

**BIG GOTCHA** 
`karma-typescript` never compiles the `source` into `ast` after transformers are run. 
As can be seen in the code, it expects transformers to provide their own AST:
https://github.com/monounity/karma-typescript/blob/master/src/bundler/transformer.ts#L89

So in order for the new option to effectively work, transformer has to do this:
```js
content.js.ast = acorn.parse(content.source, content.config.bundlerOptions.acornOptions);
```
This is not user-friendly but I did not want to change the current code expectations.
I think a good compromise would be to do the above call automatically when `transpile == true` and `content.js.ast == undefined`.

This was tested on my project and works, if you mind the gotcha.

erikbarke commented on src/bundler/transformer.ts:
```diff
@@ -83,12 +83,13 @@ export class Transformer {
         async.eachSeries(transforms, (transform: Transform, onTransformApplied: ErrorCallback<Error>) => {
             process.nextTick(() => {
-                transform(context, (error: Error, dirty: boolean) => {
+                transform(context, (error: Error, dirty: boolean, transpile: boolean = false) => {
```
We need a new parameter here, `transpile` indicates whether the source should be transpiled again or not, we can't have a parameter that sometimes changes what it means. https://github.com/jods4/karma-typescript/blob/master/src/api/transforms.ts#L7

This new parameter (`isScript` or similar) could be used together with the `transpile` parameter to check if an `ast` should be created or not.

jods4 replied:
Sorry I'm not sure I understood what you mean? Do you want a fourth parameter?

erikbarke replied:
Yeah, a fourth parameter: the api is shared between Typescript and Javascript transforms and the `transpile` parameter is used to determine whether the source should be transpiled again or not.

Maybe it would be better to use just three parameters and add two new interfaces to the api:

```js
export interface JsTransformCallback {
    (error: Error, dirty: boolean, transformedScript?: boolean): void;
};

export interface JsTransform {
    (context: TransformContext, callback: JsTransformCallback): void;
    initialize?: TransformInitialize;
};
```

And then rename the third parameter in the js transform callback to `transformedScript: boolean = false`.

What I'm looking for here is keeping the transforms api as simple an uncluttered as possible while keeping the ts and js transforms as separated as possible, it's a bit of a balancing act, I know ☺️

The js transform would look something like this:
```js
async.eachSeries(transforms, (transform: JsTransform, onTransformApplied: ErrorCallback<Error>) => {
    process.nextTick(() => {
        transform(context, (error: Error, dirty: boolean, transformedScript: boolean = false) => {
            this.handleError(error, transform, context);
            if (dirty) {
                bundleItem.ast = context.js.ast;
                bundleItem.source = context.source;
            }
            bundleItem.transformedScript = transformedScript;
            onTransformApplied();
        });
    });
}, onTransformsApplied);
```

I'll document this when we're done 🚀

jods4 replied:
I tried that but it doesn't work well because then you have `TsTransform` and `JsTransform` that are different, yet they "unify" in `configuration.bundler.transforms`.
I'd say it's best to somehow go with your first idea and keep the transforms identical.

jods4 commented:
@erikbarke What if I modify this PR and instead I add `context.js.isScript = false`. If any transform in the chain sets it to `true` its dependencies should be processed.
For ease of use I would also automatically parse AST if `dirty && isScript && !ast`.

Being a new field, it's all backward compatible.
Being on `js` it keeps the common transform apis identical for TS and JS.

Plus having lots of parameters doesn't make for very useable apis. Using an object makes it easier to consume.

erikbarke commented:
>Using an object makes it easier to consume. 

Yes, that's it! *smacks forehead*

The callback should accept booleans or an object:

`(error: Error, dirty: boolean, transpile: boolean = true)`

or

`(error: Error, result: TransformResult)`

The callback result interface:

`
export interface TransformResult {
    (dirty: boolean, transpile: boolean, transformedScript: boolean): void;
};
`

This way we still support ye olde api, letting people use the simple `dirty` flag if they want to or use a more complex result object to fine tune the transform behavior 😃

jods4 commented:
ok... you're ok with the fact that TS and JS transforms have the same sig but mostly work on different settings?
TS ignores `transformedScript` and JS ignores `transpile`?

If so I'll update the PR

erikbarke commented:
I admit that having the ts and js transforms behind the same api was a bad idea, hindsight is 20/20...
The first version of the api had just two parameters `cb(error, dirty)` but since then the api has grown to include more and more stuff and splitting the transforms into two separate lists now would be a massive breaking change 😞 So yeah, I'm ok (kind of...) with the transforms having identical signatures...

jods4 commented:
I have changed the PR according to discussion:
- Existing signatures are unchanged;
- New sig that takes an object;
- Added code to automatically compile AST if it is `undefined` and `dirty == true` and `transformedScript == true`. Makes it more intuitive and much easier to use.

erikbarke commented:
Great, I'll add this to the documentation!

jods4 commented:
Any chance to get a release? It will really the tests in my project :)

erikbarke commented:
3.0.9 is on npm now :)

jods4 commented:
thanks ❤️
